File: lua/cell.lua
Cell = {}

function Cell:new(x, y, alive)
  self.__index = self
  local cell = setmetatable({
    x = x,
    y = y,
    alive = (alive or false),
    next_state = nil,
    neighbours = {},
  }, self)

  return cell
end

function Cell:to_char()
  return self.alive and 'o' or ' '
end

function Cell:alive_neighbours()
  -- The following is slower
  -- local alive_neighbours = 0
  -- for index,neighbour in pairs(self.neighbours) do
  --   if neighbour.alive then
  --     alive_neighbours = (alive_neighbours + 1)
  --   end
  -- end
  -- return alive_neighbours

  -- The following is the fastest
  local alive_neighbours = 0
  for i = 1, #self.neighbours do
    local neighbour = self.neighbours[i]
    if neighbour.alive then
      alive_neighbours = (alive_neighbours + 1)
    end
  end
  return alive_neighbours
end

File: java/Cell.java
import java.util.ArrayList;
import java.util.List;
// import java.util.stream.Collectors;

public class Cell {
  public final int x;
  public final int y;
  public boolean alive;
  public Boolean nextState;
  public final List<Cell> neighbours;

  public Cell(int x, int y, boolean alive) {
    this.x = x;
    this.y = y;
    this.alive = alive;
    this.nextState = null;
    this.neighbours = new ArrayList<>();
  }

  public char toChar() {
    return this.alive ? 'o' : ' ';
  }

  public int aliveNeighbours() {
    // The following is slower
    // return this.neighbours.stream().
    //   filter(neighbour -> neighbour.alive).
    //   collect(Collectors.toList()).
    //   size();

    // The following is slower
    // var aliveNeighbours = 0;
    // for (var neighbour : this.neighbours) {
    //   if (neighbour.alive) {
    //     aliveNeighbours++;
    //   }
    // }
    // return aliveNeighbours;

    // The following is the fastest
    var aliveNeighbours = 0;
    var count = this.neighbours.size();
    for (var i = 0; i < count; i++) {
      var neighbour = this.neighbours.get(i);
      if (neighbour.alive) {
        aliveNeighbours++;
      }
    }
    return aliveNeighbours;
  }
}
